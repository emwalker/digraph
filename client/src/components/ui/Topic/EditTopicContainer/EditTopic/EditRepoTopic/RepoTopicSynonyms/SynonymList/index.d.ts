import React from 'react';
import { SynonymType } from 'components/types';
type Props = {
    canUpdate: boolean;
    onDelete: (position: number) => void;
    onUpdate: (synonyms: SynonymType[]) => void;
    synonyms: readonly SynonymType[];
};
export default function SynonymList({ canUpdate, onDelete, onUpdate, synonyms }: Props): React.JSX.Element;
export {};
