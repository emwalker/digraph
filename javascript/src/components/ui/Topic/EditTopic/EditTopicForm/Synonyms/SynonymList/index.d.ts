import React, { Component } from 'react';
import { Synonyms_topic$data as TopicType } from '__generated__/Synonyms_topic.graphql';
declare type RepoTopicType = TopicType['repoTopics'][0];
declare type SynonymType = RepoTopicType['synonyms'][number];
declare type Props = {
    canUpdate: boolean;
    onDelete: Function;
    onUpdate: Function;
    synonyms: readonly SynonymType[];
};
declare class SynonymList extends Component<Props> {
    onSortEnd: ({ oldIndex, newIndex }: {
        oldIndex: number;
        newIndex: number;
    }) => void;
    get canSort(): boolean;
    deleteFn: () => Function | null;
    renderReadonlyList: () => JSX.Element[];
    renderUpdatableList: () => JSX.Element;
    render: () => JSX.Element | JSX.Element[];
}
export declare const UnwrappedSynonymList: React.ComponentClass<import("react-sortable-hoc").SortableContainerProps, any>;
export default SynonymList;
