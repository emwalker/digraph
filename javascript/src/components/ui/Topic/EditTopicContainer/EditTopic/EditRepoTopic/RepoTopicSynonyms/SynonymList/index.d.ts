import { SynonymType } from 'components/types';
declare type Props = {
    canUpdate: boolean;
    onDelete: (position: number) => void;
    onUpdate: (synonyms: SynonymType[]) => void;
    synonyms: readonly SynonymType[];
};
export default function SynonymList({ canUpdate, onDelete, onUpdate, synonyms }: Props): JSX.Element;
export {};
