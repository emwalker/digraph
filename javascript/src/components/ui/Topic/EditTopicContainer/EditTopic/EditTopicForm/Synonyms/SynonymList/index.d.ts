import React, { Component } from 'react';
import { Synonyms_topic as TopicType } from '__generated__/Synonyms_topic.graphql';
declare type TopicDetailType = TopicType['details'][0];
declare type SynonymType = TopicDetailType['synonyms'][number];
declare type Props = {
    canUpdate: boolean;
    onDelete: Function;
    onUpdate: Function;
    synonyms: readonly SynonymType[];
};
declare class SynonymList extends Component<Props> {
    onSortEnd: ({ oldIndex, newIndex }: {
        oldIndex: number;
        newIndex: number;
    }) => void;
    get canSort(): boolean;
    deleteFn: () => Function | null;
    renderReadonlyList: () => JSX.Element[];
    renderUpdatableList: () => JSX.Element;
    render: () => JSX.Element | JSX.Element[];
}
export declare const UnwrappedSynonymList: React.ComponentClass<import("react-sortable-hoc").SortableContainerProps, any>;
export default SynonymList;
