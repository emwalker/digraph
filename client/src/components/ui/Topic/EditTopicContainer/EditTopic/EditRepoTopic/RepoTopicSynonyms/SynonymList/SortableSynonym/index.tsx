import React, { useCallback } from 'react'
import { useSortable } from '@dnd-kit/sortable'
import { CSS } from '@dnd-kit/utilities'
import { GoThreeBars, GoTrashcan } from 'react-icons/go'

import { SynonymType } from 'components/types'

type Props = {
  id: number,
  synonym: SynonymType,
  onDelete?: ((position: number) => void) | null,
}

const DragHandle = (props: any) => (
  <GoThreeBars
    className="synonym-drag-handle"
    style={{ width: '20px', height: '17px', color: '#d1d5da' }}
    {...props}
  />
)

export default function SortableItem({ id, synonym, onDelete }: Props) {
  const {
    attributes,
    listeners,
    setNodeRef,
    transform,
  } = useSortable({ id })
  const style = { transform: CSS.Transform.toString(transform) }

  const onClick = useCallback(() => {
    if (!onDelete || !id) return
    onDelete(id)
  }, [onDelete, id])

  return (
    <li
      ref={setNodeRef}
      style={style}
      className="Box-row clearfix css-truncate p-2 d-flex"
    >
      <DragHandle {...listeners} {...attributes} />

      <div className="col-10">{synonym.name}</div>
      <div className="col-1">
        { synonym.locale }
      </div>

       <div className="col-1 remove-synonym">
        {onDelete && (
          <span tabIndex={0} role="button" onClick={onClick}>
            <GoTrashcan />
          </span>
        )}
      </div>
    </li>
  )
}
