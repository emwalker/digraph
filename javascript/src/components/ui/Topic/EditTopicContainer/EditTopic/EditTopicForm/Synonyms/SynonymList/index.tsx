import React, { Component } from 'react'
import {
  SortableContainer,
  SortableElement,
  SortableHandle,
} from 'react-sortable-hoc'
import arrayMove from 'array-move'
import { GoThreeBars } from 'react-icons/go'

import { Synonyms_topic as TopicType } from '__generated__/Synonyms_topic.graphql'
import Synonym from '../Synonym'
import copySynonyms from '../copySynonyms'

type SynonymType = TopicType['synonyms'][number]

type Props = {
  canUpdate: boolean,
  onDelete: Function,
  onUpdate: Function,
  synonyms: readonly SynonymType[],
}

const DragHandle = SortableHandle(() => (
  <GoThreeBars
    className="synonym-drag-handle"
    style={{ width: '20px', height: '20px', color: '#d1d5da' }}
  />
))

const SortableSynonym = SortableElement((props: any) => <Synonym {...props} />)

type ContainerProps = {
  items: readonly SynonymType[],
  onDelete: Function | null,
}

const SortableList = SortableContainer(({ onDelete, items }: ContainerProps) => (
  <div>
    {
      items.map((synonym, index) => (
        <SortableSynonym
          // @ts-ignore
          dragHandle={<DragHandle />}
          index={index}
          key={synonym.name}
          onDelete={onDelete}
          position={index}
          synonym={synonym}
        />
      ))
    }
  </div>
))

class SynonymList extends Component<Props> {
  onSortEnd = ({ oldIndex, newIndex }: { oldIndex: number, newIndex: number }) => {
    if (!this.props.canUpdate) return

    const synonyms = arrayMove(this.props.synonyms, oldIndex, newIndex)
    this.props.onUpdate(copySynonyms(synonyms))
  }

  get canSort(): boolean {
    return this.props.canUpdate && this.props.synonyms.length > 1
  }

  deleteFn = () => (
    this.props.canUpdate
      ? this.props.onDelete
      : null
  )

  renderReadonlyList = () => (
    this.props.synonyms.map((value) => (
      <Synonym key={value.name} synonym={value} />
    ))
  )

  renderUpdatableList = () => (
    <SortableList
      // @ts-ignore
      items={this.props.synonyms}
      lockAxis="y"
      onDelete={this.deleteFn()}
      onSortEnd={this.onSortEnd}
      useDragHandle
    />
  )

  render = () => (
    this.canSort
      ? this.renderUpdatableList()
      : this.renderReadonlyList()
  )
}

export const UnwrappedSynonymList = SortableList

export default SynonymList
