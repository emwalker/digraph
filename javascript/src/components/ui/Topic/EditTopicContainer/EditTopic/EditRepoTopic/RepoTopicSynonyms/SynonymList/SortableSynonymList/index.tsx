import React, { useCallback } from 'react'
import {
  DndContext, closestCenter, KeyboardSensor, PointerSensor, useSensor, useSensors, DragEndEvent,
} from '@dnd-kit/core'
import {
  arrayMove, SortableContext, sortableKeyboardCoordinates, verticalListSortingStrategy,
} from '@dnd-kit/sortable'

import { SynonymType } from 'components/types'
import SortableSynonym from '../SortableSynonym'

type Props = {
  synonyms: readonly SynonymType[],
  onDelete: ((position: number) => void) | null,
  onUpdate: (synonyms: SynonymType[]) => void,
}

export default function SortableSynonymList({ synonyms, onUpdate, onDelete }: Props) {
  const sensors = useSensors(
    useSensor(PointerSensor),
    useSensor(KeyboardSensor, {
      coordinateGetter: sortableKeyboardCoordinates,
    }),
  )

  const sortedIds = synonyms.map((_, index) => index)

  const handleDragEnd = useCallback((event: DragEndEvent) => {
    const { active, over } = event

    if (!over) return
    if (active.id === over.id) return

    const oldIndex = sortedIds.indexOf(active.id as number)
    const newIndex = sortedIds.indexOf(over.id as number)
    const newIds = arrayMove(sortedIds, oldIndex, newIndex)
    onUpdate(newIds.map((id) => synonyms[id]))
  }, [sortedIds, synonyms, onUpdate, arrayMove])

  return (
    <DndContext
      sensors={sensors}
      collisionDetection={closestCenter}
      onDragEnd={handleDragEnd}
    >
      <SortableContext
        items={sortedIds}
        strategy={verticalListSortingStrategy}
      >
        {sortedIds.map((id) => (
          <SortableSynonym key={id} id={id} synonym={synonyms[id]} onDelete={onDelete} />
        ))}
      </SortableContext>
    </DndContext>
  )
}
