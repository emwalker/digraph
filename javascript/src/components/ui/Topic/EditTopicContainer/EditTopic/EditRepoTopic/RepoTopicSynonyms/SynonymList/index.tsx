import React from 'react'

import { SynonymType } from 'components/types'
import SortableSynonym from './SortableSynonym'
import SortableSynonymList from './SortableSynonymList'

type Props = {
  canUpdate: boolean,
  onDelete: (position: number) => void,
  onUpdate: (synonyms: SynonymType[]) => void,
  synonyms: readonly SynonymType[],
}

export default function SynonymList({ canUpdate, onDelete, onUpdate, synonyms }: Props) {
  const canSort = canUpdate && synonyms.length > 1

  if (!canSort) {
    return (
      <div data-testid="synonym-list">
        {synonyms.map((value, index) => (
          <SortableSynonym key={value.name} id={index} synonym={value} />
        ))}
      </div>
    )
  }

  return (
    <div data-testid="synonym-list">
      <SortableSynonymList
        synonyms={synonyms}
        onDelete={onDelete}
        onUpdate={onUpdate}
      />
    </div>
  )
}
