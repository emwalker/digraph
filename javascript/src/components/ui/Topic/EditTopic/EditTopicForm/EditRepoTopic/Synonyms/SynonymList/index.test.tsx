import React from 'react'
import { shallow } from 'enzyme'

import SynonymList from './index'

const synonym = {
  name: 'Gnusto',
  id: '1234',
}

const props = {
  canUpdate: true,
  synonyms: [synonym],
  onDelete: jest.fn(),
  topic: {
    name: 'Gnusto',
    id: '1234',
  },
}

describe('<SynonymList />', () => {
  // @ts-expect-error
  const wrapper = shallow(<SynonymList {...props} />)

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
