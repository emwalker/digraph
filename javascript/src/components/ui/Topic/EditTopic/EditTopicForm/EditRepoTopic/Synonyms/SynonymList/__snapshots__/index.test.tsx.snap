// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SynonymList /> renders 1`] = `
<Synonym
  key="Gnusto"
  synonym={
    Object {
      "id": "1234",
      "name": "Gnusto",
    }
  }
/>
`;
